Fix inconsistent casting of Bool type in IN operator

ashrithb opened the pull request:
```sql
SELECT CAST(1, 'Bool') IN (10);
0
SELECT CAST(1, 'Bool') IN (255);
0
SELECT CAST(1, 'Bool') IN (256);
1 (should be 0)
SELECT CAST(1, 'Bool') IN (10000);
1 (should be 0)
```

### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The `IN` operator now uses exact value semantics for `Bool` type: only `0` and `1` values in the set match `Bool` values. Previously, numeric values greater than `255` in the `IN` set were incorrectly clamped to true when compared against `Bool`, so `SELECT CAST(1, 'Bool') IN (256)` returned 1. Now it correctly returns `0`. Closes https://github.com/ClickHouse/ClickHouse/issues/92980.

<!-- ch-version-info:start -->
### Version info
- Merged into: `26.4.1.415`
<!-- ch-version-info:end -->

ashrithb committed:
Fix inconsistent casting of Bool type in IN operator (#92980)

The Bool type in ClickHouse is implemented as UInt8 with a custom name.
When converting values for the IN operator, the equals() check between
UInt8 and Bool types returns true (same underlying typeid), causing
values like 10 or 255 to be stored as-is instead of being converted
to 0/1 (false/true).

This fix adds a special case in both convertFieldToType() and
convertFieldToTypeImpl() to detect when the target type is Bool
but the source type is not Bool (e.g., plain UInt8), and proceed
with proper Bool conversion instead of short-circuiting.

Before fix:
  SELECT CAST(1, 'Bool') IN (10);   -- returned 0 (incorrect)
  SELECT CAST(1, 'Bool') IN (255);  -- returned 0 (incorrect)
  SELECT CAST(1, 'Bool') IN (256);  -- returned 1 (correct)
  SELECT CAST(1, 'Bool') IN (10000);-- returned 1 (correct)

After fix:
  All non-zero values correctly convert to Bool true (1)
  Zero correctly converts to Bool false (0)

nihalzp reviewed (commented):
Thanks for working on the issue! Please add a test.

ashrithb committed:
Add test for Bool IN operator casting fix (#92980)

Test covers:
- Basic tests from the issue: values 10, 255, 256, 10000 with Bool true
- Zero value correctly matches Bool false
- Bool false with various numeric literals
- Multiple values in IN clause
- Negative numbers (convert to true)
- Materialized values to test non-constant path

ashrithb commented:
Added in a test @nihalzp!

ashrithb committed:
Trigger CI re-run to verify flaky test failures are unrelated to this PR

nihalzp reviewed (commented):
The changes fixed the bugs from the issue! I noticed there are some other cases of `Bool` conversion that should work. For example,

```sql
SELECT CAST(1, 'Nullable(Bool)') IN (toNullable(10));
0
SELECT CAST(1, 'Nullable(Bool)') IN (toNullable(255));
0
SELECT CAST(1, 'Nullable(Bool)') IN (toNullable(256));
1
SELECT CAST(1, 'Nullable(Bool)') IN (toNullable(1000));
1
```
I think it might be best if we fix them with this PR. 

Additionally, the PR has additionally fixed the following kind of queries:

```sql
SELECT toLowCardinality(CAST(1, 'Bool')) IN (10);
SELECT toLowCardinality(CAST(1, 'Bool')) IN (255);
```

Let's add these to tests as well.

ashrithb committed:
added in new fixes for the tests in nihal's comment + added those to test suite in 03778

ashrithb committed:
Merge branch 'master' into fix/bool-in-casting-92980

nihalzp committed:
Merge branch 'master' into fix/bool-in-casting-92980

nihalzp committed:
Merge remote-tracking branch 'origin/master' into bool-in-casting-92980

nihalzp committed:
Revert changes

nihalzp committed:
Fix and improve the test

nihalzp committed:
Add edge cases to test

nihalzp committed:
Fix semantics
